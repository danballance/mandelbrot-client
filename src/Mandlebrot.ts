export interface MandlebrotStruct {
    xmin: number;
    xmax: number;
    ymin: number;
    ymax: number;
    r: number;
    rmin: number;
    rmax: number;
    j: number;
    jmin: number;
    jmax: number;
    magnification: number;
    iterations: number;
    dpi: number;
}

export interface MandlebrotInterface {
    init(size: number, r: number, j: number, magnification: number, iterations: number, dpi: number): MandlebrotStruct;
    updateR(m: MandlebrotStruct, r: number, magnification: number): MandlebrotStruct;
    updateJ(m: MandlebrotStruct, j: number, magnification: number): MandlebrotStruct;
    getNewR(m: MandlebrotStruct, y: number): number;
    getNewJ(m: MandlebrotStruct, x: number): number;
    imageClicked(m: MandlebrotStruct, x: number, y: number): MandlebrotStruct;
    zoom(m: MandlebrotStruct, multiplier: number): MandlebrotStruct; 
}

export const Mandlebrot = (): MandlebrotInterface => {
    function init(
        size: number,
        r: number,
        j: number,
        magnification: number,
        iterations: number,
        dpi: number
    ): MandlebrotStruct {
        return {
            xmin: 0.00,
            xmax: 0.00 + size,
            ymin: 0.00,
            ymax: 0.00 + size,
            r: r,
            rmin: r - (magnification / 2),
            rmax: r + (magnification / 2),
            j: j,
            jmin: j - (magnification / 2),
            jmax: j + (magnification / 2),
            magnification: magnification,
            iterations: iterations,
            dpi
        }
    }

    function updateR(m: MandlebrotStruct, r: number, magnification: number): MandlebrotStruct {
        return Object.assign({}, m, {
            r: r,
            rmin: r - (magnification / 2),
            rmax: +r + +(magnification / 2),
            magnification: magnification 
        });
    }

    function updateJ(m: MandlebrotStruct, j: number, magnification: number): MandlebrotStruct {
        return Object.assign({}, m, {
            j: j,
            jmin: j - (magnification / 2),
            jmax: +j + +(magnification / 2),
            magnification: magnification 
        });
    }

    function getNewR(m: MandlebrotStruct, y: number): number {
        return m.rmin + (((m.ymax - y) - m.ymin) / (m.ymax - m.ymin)) * m.magnification;
    }

    function getNewJ(m: MandlebrotStruct, x: number): number {
        return m.jmax - ((x - m.xmin) / (m.xmax - m.xmin)) * m.magnification;
    }

    function zoom(m: MandlebrotStruct, multiplier: number): MandlebrotStruct {
        const newMagnification = m.magnification * multiplier;
        const m2 = updateR(m, m.r, newMagnification);
        return updateJ(m2, m2.j, newMagnification);
    }

    function imageClicked(m: MandlebrotStruct, x: number, y: number): MandlebrotStruct {
        const m2 = updateR(m, getNewR(m, y), m.magnification);
        const m3 = updateJ(m2, getNewJ(m2, x), m2.magnification);
        return zoom(m3, 0.5);
    }

    return Object.freeze({
        init,
        updateR,
        updateJ,
        getNewR,
        getNewJ,
        imageClicked,
        zoom
    });
}
