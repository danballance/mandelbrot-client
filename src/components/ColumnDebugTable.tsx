import * as React from "react";
import { Grid, Table } from 'semantic-ui-react'
import { MandlebrotStruct } from "../Mandlebrot";

interface Props {
    mandlebrotStruct: MandlebrotStruct;
    url: string;
    imageLoaded: boolean;
    mouseX: number | null;
    mouseY: number | null;
    newR: number | null;
    newJ: number | null;
    size: number;
}

export const ColumnDebugTable: React.FC<Props> =
    (props: Props): JSX.Element => (
        <Grid.Column only="computer" computer={8} textAlign="left">
            <Table inverted fixed striped compact="very" className="debug">
                <Table.Header>
                    <Table.Row>
                        <Table.HeaderCell>Property</Table.HeaderCell>
                        <Table.HeaderCell>Value</Table.HeaderCell>
                    </Table.Row>
                </Table.Header>
                <Table.Body>
                    <Table.Row>
                        <Table.Cell>Mouse X</Table.Cell>
                        <Table.Cell id="mouseX">{props.mouseX}px</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>Mouse Y</Table.Cell>
                        <Table.Cell id="mouseY">{props.mouseY}px</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>New R</Table.Cell>
                        <Table.Cell id="newR">{props.newR}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>New J</Table.Cell>
                        <Table.Cell id="newJ">{props.newJ}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>Zoom</Table.Cell>
                        <Table.Cell id="magnification">
                            {props.mandlebrotStruct.magnification}
                        </Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>Size</Table.Cell>
                        <Table.Cell id="size">{props.size}px</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>Real (R)</Table.Cell>
                        <Table.Cell id="r">{props.mandlebrotStruct.r}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>R Min</Table.Cell>
                        <Table.Cell id="rmin">{props.mandlebrotStruct.rmin}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>R Max</Table.Cell>
                        <Table.Cell id="rmax">{props.mandlebrotStruct.rmax}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>Imaginary (J)</Table.Cell>
                        <Table.Cell id="j">{props.mandlebrotStruct.j}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>J Min</Table.Cell>
                        <Table.Cell id="jmin">{props.mandlebrotStruct.jmin}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>J Max</Table.Cell>
                        <Table.Cell id="jmax">{props.mandlebrotStruct.jmax}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>Iterations</Table.Cell>
                        <Table.Cell id="iterations">{props.mandlebrotStruct.iterations}</Table.Cell>
                    </Table.Row>
                    <Table.Row>
                        <Table.Cell>DPI</Table.Cell>
                        <Table.Cell id="dpi">{props.mandlebrotStruct.dpi}</Table.Cell>
                    </Table.Row>
                </Table.Body>
            </Table>
        </Grid.Column>
    )