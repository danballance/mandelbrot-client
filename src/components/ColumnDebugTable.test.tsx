import React from 'react';
import { shallow } from 'enzyme'
import { ColumnDebugTable } from './ColumnDebugTable'
import { Mandlebrot } from '../Mandlebrot'

const mandlebrotStruct = Mandlebrot().init(400, -0.5, 0.0, 2.5, 100, 200);
const columnDebugTable = shallow(<ColumnDebugTable
    mandlebrotStruct={mandlebrotStruct}
    url='https://localhost/api'
    imageLoaded={false}
    mouseX={320}
    mouseY={120}
    newR={0.00}
    newJ={-0.75}
    size={400}
/>);

describe('ColumnDebugTable', (): void => {
    it('sets the mouseX cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#mouseX').render().text()
        ).toBe('320px');
    })

    it('sets the mouseY cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#mouseY').render().text()
        ).toBe('120px');
    })

    it('sets the newR cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#newR').render().text()
        ).toBe('0');
    })

    it('sets the newJ cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#newJ').render().text()
        ).toBe('-0.75');
    })

    it('sets the magnification cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#magnification').render().text()
        ).toBe('2.5');
    })

    it('sets the size cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#size').render().text()
        ).toBe('400px');
    })

    it('sets the r cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#r').render().text()
        ).toBe(mandlebrotStruct.r.toString());
    })

    it('sets the rmin cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#rmin').render().text()
        ).toBe(mandlebrotStruct.rmin.toString());
    })

    it('sets the rmax cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#rmax').render().text()
        ).toBe(mandlebrotStruct.rmax.toString());
    })

    it('sets the j cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#j').render().text()
        ).toBe(mandlebrotStruct.j.toString());
    })

    it('sets the jmin cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#jmin').render().text()
        ).toBe(mandlebrotStruct.jmin.toString());
    })

    it('sets the jmax cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#jmax').render().text()
        ).toBe(mandlebrotStruct.jmax.toString());
    })

    it('sets the iterations cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#iterations').render().text()
        ).toBe('100');
    })

    it('sets the dpi cell correctly', (): void => {
        expect(
            columnDebugTable.find('TableCell#dpi').render().text()
        ).toBe('200');
    })

    it('renders correctly', (): void => {
        expect(columnDebugTable).toMatchSnapshot();
    })
});
