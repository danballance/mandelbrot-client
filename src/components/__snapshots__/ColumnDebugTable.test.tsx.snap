// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnDebugTable renders correctly 1`] = `ShallowWrapper {}`;
